components/x-interaction/src/InteractionClass.jsx

```diff
@@ -11,32 +11,32 @@ export class InteractionClass extends Component {
 			inFlight: 0,
 		};
 
-		this.actions = mapValues(props.actions, (func) => (...args) => {
+		this.actions = mapValues(props.actions, (func) => async (...args) => {
 			// mark as loading one microtask later. if the action is synchronous then
 			// setting loading back to false will happen in the same microtask and no
 			// additional render will be scheduled.
 			Promise.resolve().then(() => {
 				this.setState(({ inFlight }) => ({ inFlight: inFlight + 1 }));
 			});
 
-			return Promise.resolve(func(...args)).then((next) => {
-				const updater = typeof next === 'function'
-					? ({state}) => ({state: Object.assign(
-						state,
-						next(Object.assign(
-							{},
-							props.initialState,
-							state
-						))
-					)})
-					: ({state}) => ({state: Object.assign(state, next)});
-
-				return new Promise(resolve =>
-					this.setState(updater, () => (
-						this.setState(({ inFlight }) => ({ inFlight: inFlight - 1 }), resolve)
-					))
-				);
-			});
+			const stateUpdate = await Promise.resolve(func(...args));
+
+			const nextState = typeof stateUpdate === 'function'
+				? Object.assign(
+					this.state.state,
+					await Promise.resolve(stateUpdate(Object.assign(
+						{},
+						props.initialState,
+						this.state.state
+					)))
+				)
+				: Object.assign(this.state.state, stateUpdate);
```

Comment:
Do you reckon it is worth micro-optimising these `Object.assign()` calls by using a faster merge function?

Reply:
do you have a preferred merge? i'm not terribly concerned about performance here as a) they're async, so they won't be blocking UI, and b) if somebody's triggering actions in a tight loop how about instead of doing that they don't